File: go/MaxDepthBinaryTree/MaxDepthBinaryTree.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Time -> 0(N) Space -> 0(log(N))
func maxDepth(root *TreeNode) int {
	return findDepth(root, 0)
}

func findDepth(root *TreeNode, depth int) int {
	if root == nil {
		return depth
	}

	newDepth := depth + 1
	leftDepth := findDepth(root.Left, newDepth)
	rightDepth := findDepth(root.Right, newDepth)

	return max(leftDepth, rightDepth)
}

func main() {
	left := &TreeNode{Val: 9}
	right := &TreeNode{Val: 20}
	root := &TreeNode{Val: 3, Left: left, Right: right}
	fmt.Println(maxDepth(root))
}

File: python/maxDepthBinaryTree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


# Time -> 0(N) Space -> 0(log(N))
def maxDepth(root) -> int:
    return findDepth(root, 0)


def findDepth(root, depth) -> int:
    if root is None:
        return depth

    newDepth = depth + 1
    leftDepth = findDepth(root.left, newDepth)
    rightDepth = findDepth(root.right, newDepth)

    return max(leftDepth, rightDepth)


right = TreeNode(val=20)
left = TreeNode(val=9)
root = TreeNode(val=3, left=left, right=right)
print(maxDepth(root))
